### PR TITLE
Replacement Apply and Resample

### PR DESCRIPTION
**Pull request by atbenmurray**
Signed-off-by: Ben Murray <ben.murray@gmail.com>

### Description

This is part of the work towards #4855. It adds:
 - a lazy `apply` method
 - A transform-like wrapper for `apply` called `Apply`
 ~- `MetaMatrix` and related functionality to represent abstracted grid and matrix transforms with metadata~
 - A universal `resample` function that can be used to apply grid / matrix transforms
 ~- Functional spatial and croppad implementations that define but don't apply transforms~

### Types of changes

- [x] Non-breaking change (fix or new feature that would not break existing functionality).
- [ ] Breaking change (fix or new feature that would cause existing functionality to change).
- [x] New tests added to cover the changes.
- [ ] Integration tests passed locally by running `./runtests.sh -f -u --net --coverage`.
- [x] Quick tests passed locally by running `./runtests.sh --quick --unittests  --disttests`.
- [x] In-line docstrings updated.
- [ ] Documentation updated, tested `make html` command in the `docs/` folder.

### CODE REVIEW

File: monai/transforms/lazy/functional.py [hunk 1]
```diff
@@ -0,0 +1,201 @@
...
+    cumulative_matrix_ = matrix_from_matrix_container(cumulative_matrix)
+
+    # print(f"applying with cumulative matrix\n {cumulative_matrix_}")
+    a = Affine(norm_coords=False, affine=cumulative_matrix_, spatial_size=cur_shape[1:], normalized=False, **kwargs)
+    data, tx = a(img=data)
+    if isinstance(data, MetaTensor):
+        data.clear_pending_transforms()
+        for p in pending_:
+            data.affine = p.matrix.data
+            data.push_applied_operation(p)
```
Comment (wyli) [inline monai/transforms/lazy/functional.py:199]: thanks for the updates, could you create an array transform example based on this? the previous #4922's transform seems to be out of date (https://github.com/Project-MONAI/MONAI/blob/62f8172f020c3c8b6ac182b6b4ca5c3805f9877b/monai/transforms/atmostonce/array.py)

**Comment by wyli** [comment]
thanks, I think this PR introduces quite a few APIs at a time and some of them (`MatrixFactory`/`MetaMatrix`) need more discussions... is it possible to make it smaller, for example, only including `croppad`, `apply`, and an example of `SpatialCrop` transform, so that we can create a working example of lazy cropping, then finalise the `functional` and `apply` API?  cc @ericspod @rijobro @Nic-Ma

**Comment by atbenmurray** [comment]
> thanks, I think this PR introduces quite a few APIs at a time and some of them (`MatrixFactory`/`MetaMatrix`) need more discussions... is it possible to make it smaller, for example, only including `croppad`, `apply`, and an example of `SpatialCrop` transform, so that we can create a working example of lazy cropping, then finalise the `functional` and `apply` API? cc @ericspod @rijobro @Nic-Ma

The problem is that MatrixFactory/MetaMatrix are the classes by which a lot of the internal stuff for lazy resampling is done. They could certainly change before the PR goes in, but that would necessitate some plumbing to happen and I don't know if I'll have time before I am on holiday. The key thing here at the moment is that this PR doesn't change any existing behaviour so if we want to refactor we certainly can.

What they do:
 - MatrixFactory is an attempt to wrap up the various affine operations in a way that it can be simply got from a tensor and used in an OO way.
 - MetaTensor is the transformation + metadata that accompanies it. It contains either a Matrix or a Grid object, depending

In theory, one or both could go, but it might make everything somewhat verbose. Neither are supposed to be user-facing

Might be a good topic for tomorrow, but I am trying to get the PRs as populated as possible for tonight.

File: monai/transforms/meta_matrix.py [hunk 2]
```diff
@@ -0,0 +1,292 @@
...
+
+    def __rmatmul__(self, other):
+        if isinstance(other, MetaMatrix):
+            other_ = other.matrix
+        else:
+            other_ = other
+        return MetaMatrix(other_ @ self.matrix)
+
+
+def matmul(
```
Comment (wyli) [inline monai/transforms/meta_matrix.py:177]: this util almost addresses https://github.com/Project-MONAI/MONAI/issues/4604 if the interface is simpler and considers a batch of transforms... any interests in tackling that ticket within this implementation? @atbenmurray

Comment (atbenmurray) [reply monai/transforms/meta_matrix.py:177]: @wyli sorry, access has been patchy. Let me take a look

Comment (atbenmurray) [reply monai/transforms/meta_matrix.py:177]: @wyli Yes, I think so. I haven't understood yet how to do it so if there are well defined methods that is great!

Comment (atbenmurray) [reply monai/transforms/meta_matrix.py:177]: @wyli Hmm, it does appear to be the same principle. Well, at least we are pytorch and they are tensorflow. I did look across a number of libraries but I focused on the pytorch-based ones of course

Comment (atbenmurray) [reply monai/transforms/meta_matrix.py:177]: @wyli hmm, it seems to be more focused on providing these capabilities for registration based deep learning, as opposed to preprocessing

Comment (wyli) [reply monai/transforms/meta_matrix.py:177]: if this `matmul` could be the util to implement the key idea and shared between preprocessing and as a differentiable module , that would be nice...

File: monai/transforms/lazy/utils.py [hunk 3]
```diff
@@ -0,0 +1,123 @@
...
+        if isinstance(data, Affine):
+            return True
+        if isinstance(data, DDF):
+            return False
+        if not hasattr(data, "shape") or len(data.shape) < 2:
+            return False
+        return data.shape[-1] in (3, 4) and data.shape[-2] in (3, 4) and data.shape[-1] == data.shape[-2]
+
+
+class DDF:
```
Comment (wyli) [inline monai/transforms/lazy/utils.py:44]: `class DisplacementField`

**Review by wyli — APPROVED** [review]
@atbenmurray please make more changes if needed and let me know when you think it's ready.. cc @Nic-Ma this mainly introduces necessary APIs `monai.transforms.lazy.*`

**Pull request merged**